tsweb: flake in TestStdHandler_ConnectionClosedDuringBody

Noticed in an unrelated PR.

https://github.com/tailscale/tailscale/actions/runs/10231116493/job/28306694891?pr=13015
https://github.com/tailscale/tailscale/actions/runs/10231116493/job/28306695053?pr=13015
https://github.com/tailscale/tailscale/actions/runs/10231116493/job/28306695098?pr=13015

```
=== RUN   TestStdHandler_ConnectionClosedDuringBody
    tsweb_test.go:927: got error <nil>, want context.Canceled
    tsweb_test.go:930: got response &http.Response{Status:"200 OK", StatusCode:200, Proto:"HTTP/1.1", ProtoMajor:1, ProtoMinor:1, Header:http.Header{"Content-Type":[]string{"application/octet-stream"}, "Date":[]string{"Sat, 03 Aug 2024 21:43:17 GMT"}}, Body:(*http.bodyEOFSignal)(0x140000cc980), ContentLength:-1, TransferEncoding:[]string{"chunked"}, Close:false, Uncompressed:false, Trailer:http.Header(nil), Request:(*http.Request)(0x14000549560), TLS:(*tls.ConnectionState)(nil)}, want nil
    tsweb.go:577: {"time":"2024-08-03T14:43:17.716243-07:00","remote_addr":"127.0.0.1:57428","proto":"HTTP/1.1","host":"127.0.0.1:57427","method":"GET","request_uri":"/","user_agent":"Go-http-client/1.1","code":499,"bytes":196608,"err":"net/http: abort Handler (original code 200)"}
--- FAIL: TestStdHandler_ConnectionClosedDuringBody (0.00s)
```

cc @icio

--- a/tsweb/tsweb_test.go
+++ b/tsweb/tsweb_test.go
@@ -22,7 +22,6 @@ import (
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
-	"tailscale.com/cmd/testwrapper/flakytest"
 	"tailscale.com/metrics"
 	"tailscale.com/tstest"
 	"tailscale.com/util/httpm"
@@ -865,14 +864,14 @@ func TestStdHandler_CanceledAfterHeader(t *testing.T) {
 }
 
 func TestStdHandler_ConnectionClosedDuringBody(t *testing.T) {
-	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/13017")
 	now := time.Now()
 
-	// Start a HTTP server that returns 1MB of data.
+	// Start a HTTP server that writes back zeros until the request is abandoned.
 	// We next put a reverse-proxy in front of this server.
 	rs := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		for range 1024 {
-			w.Write(make([]byte, 1024))
+		zeroes := make([]byte, 1024)
+		for r.Context().Err() == nil {
+			w.Write(zeroes)
 		}
 	}))
 	defer rs.Close()
@@ -880,8 +879,9 @@ func TestStdHandler_ConnectionClosedDuringBody(t *testing.T) {
 	r := make(chan AccessLogRecord)
 	var e *HTTPError
 	responseStarted := make(chan struct{})
+	requestCanceled := make(chan struct{})
 
-	// Create another server which proxies our 1MB server.
+	// Create another server which proxies our zeroes server.
 	// The [httputil.ReverseProxy] will panic with [http.ErrAbortHandler] when
 	// it fails to copy the response to the client.
 	h := StdHandler(
@@ -889,10 +889,6 @@ func TestStdHandler_ConnectionClosedDuringBody(t *testing.T) {
 			(&httputil.ReverseProxy{
 				Director: func(r *http.Request) {
 					r.URL = must.Get(url.Parse(rs.URL))
-				},
-				ModifyResponse: func(r *http.Response) error {
-					close(responseStarted)
-					return nil
 				},
 			}).ServeHTTP(w, r)
 			return nil
@@ -908,7 +904,11 @@ func TestStdHandler_ConnectionClosedDuringBody(t *testing.T) {
 			},
 		},
 	)
-	s := httptest.NewServer(h)
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		close(responseStarted)
+		<-requestCanceled
+		h.ServeHTTP(w, r.WithContext(context.WithoutCancel(r.Context())))
+	}))
 	t.Cleanup(s.Close)
 
 	// Create a context which gets canceled after the handler starts processing
@@ -925,6 +925,7 @@ func TestStdHandler_ConnectionClosedDuringBody(t *testing.T) {
 		t.Fatalf("making request: %s", err)
 	}
 	res, err := http.DefaultClient.Do(req)
+	close(requestCanceled)
 	if !errors.Is(err, context.Canceled) {
 		t.Errorf("got error %v, want context.Canceled", err)
 	}

